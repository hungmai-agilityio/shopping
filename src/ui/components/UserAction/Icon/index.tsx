import { Icon } from '@/ui/components';

interface UserIconProps {
  src: string;
  alt: string;
  onClick?: () => void;
  isBadge?: boolean;
  badgeCount?: number;
}

const UserIcon = ({
  src,
  alt,
  onClick,
  isBadge = false,
  badgeCount = 0
}: UserIconProps) => {
  return (
    <Icon
      src={src}
      alt={alt}
      width={24}
      height={24}
      onClick={onClick}
      isBadge={isBadge}
      badgeCount={badgeCount}
      priority
    />
  );
};

export default UserIcon;
