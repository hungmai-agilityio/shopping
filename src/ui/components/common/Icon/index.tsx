import { clsx } from 'clsx';
import Image, { ImageProps } from 'next/image';

interface IconProps extends ImageProps {
  isBadge?: boolean;
  badgeCount?: number;
  onClick?: () => void;
  isDisabled?: boolean;
}
const Icon = ({
  src,
  alt,
  width,
  height,
  isBadge,
  badgeCount,
  onClick,
  isDisabled,
  ...props
}: IconProps) => {
  return (
    <button
      className={clsx('relative inline-flex items-center justify-center', {
        'opacity-20': isDisabled
      })}
      onClick={onClick}
      disabled={isDisabled}
    >
      <Image src={src} alt={alt} width={width} height={height} {...props} />
      {isBadge && (
        <span
          className={clsx(
            'absolute -top-2 -right-2 h-5 w-5 rounded-full bg-red-400 text-white text-xs font-bold flex items-center justify-center',
            badgeCount === 0 && 'hidden'
          )}
        >
          {badgeCount}
        </span>
      )}
    </button>
  );
};

export default Icon;
