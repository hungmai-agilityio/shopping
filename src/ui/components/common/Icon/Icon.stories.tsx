import { Meta, StoryObj } from '@storybook/react';

// Components
import Icon from '@/ui/components/common/Icon';

const meta: Meta<typeof Icon> = {
  component: Icon,
  tags: ['autodocs'],
  parameters: {
    layout: 'centered'
  },
  argTypes: {
    src: {
      description: 'Image to render'
    },

    alt: {
      description: 'Describes the image and shows if the image is corrupted'
    },

    width: {
      description: 'Width for images'
    },
    height: {
      description: 'Height for images'
    },

    isBadge: {
      description: 'Attribute for users to choose whether to show badges or not'
    },

    badgeCount: {
      description: 'The number that the badge shows'
    },
    onClick: {
      description: 'Handle the event when clicking on the button'
    }
  }
};

export default meta;

type Story = StoryObj<typeof Icon>;

export const Default: Story = {
  args: {
    src: '/user.svg',
    alt: 'Cart Icon',
    width: 24,
    height: 24,
    onClick: () => {
      alert(123);
    }
  }
};

export const HasBadge: Story = {
  args: {
    src: '/bag.svg',
    alt: 'Cart Icon',
    width: 24,
    height: 24,
    isBadge: true,
    badgeCount: 3,
    onClick: () => {
      alert(123);
    }
  }
};

export const NotAction: Story = {
  args: {
    src: '/star.svg',
    alt: 'Cart Icon',
    width: 106,
    height: 18
  }
};
