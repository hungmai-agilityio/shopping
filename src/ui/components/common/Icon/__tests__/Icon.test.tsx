import Icon from '@/ui/components/common/Icon';
import { fireEvent, render, screen } from '@testing-library/react';

describe('Icon component', () => {
  const onClickMock = jest.fn();

  const defautProps = {
    src: '/user.svg',
    alt: 'Cart Icon',
    width: 24,
    height: 24,
    onClick: onClickMock
  };

  it('renders icon with default props', () => {
    render(<Icon {...defautProps} />);

    const icon = screen.getByAltText('Cart Icon');
    expect(icon).toBeInTheDocument();
  });

  it('renders icon with badge props', () => {
    render(<Icon {...defautProps} isBadge={true} badgeCount={3} />);

    const icon = screen.getByAltText('Cart Icon');
    expect(icon).toBeInTheDocument();
    const badge = screen.getByText('3');
    expect(badge).toHaveClass('bg-red-400 text-white text-xs font-bold');
  });

  it('hides badge when badgeCount is 0', () => {
    render(<Icon {...defautProps} isBadge={true} badgeCount={0} />);

    const icon = screen.getByAltText('Cart Icon');
    expect(icon).toBeInTheDocument();

    const badge = screen.queryByText('0');
    if (badge) {
      expect(badge).toHaveClass('hidden');
    }
  });

  it('calls the onClick handler when clicked', () => {
    render(<Icon {...defautProps} />);

    const icon = screen.getByAltText('Cart Icon');
    fireEvent.click(icon);

    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  it('should be render match to snapshot', () => {
    const { container } = render(<Icon {...defautProps} />);

    expect(container).toMatchSnapshot();
  });
});
