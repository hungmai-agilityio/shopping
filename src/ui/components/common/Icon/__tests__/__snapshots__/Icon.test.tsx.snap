// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component should be render match to snapshot 1`] = `
<div>
  <button
    class="relative inline-flex items-center justify-center"
  >
    <img
      alt="Cart Icon"
      data-nimg="1"
      decoding="async"
      height="24"
      loading="lazy"
      src="/user.svg"
      style="color: transparent;"
      width="24"
    />
  </button>
</div>
`;
